### PR TITLE
Refactor password generation

### PR DESCRIPTION
**Pull request by FabianHenneke**
## :loudspeaker: Type of change

- [ ] Bugfix
- [ ] New feature
- [ ] Enhancement
- [x] Refactoring
- [ ] Dependency updates

## :scroll: Description

Refactors and improves the password generation code (phonemes, completely random & XKCD) in the following way:
- Extract all secure random number generation into a single file with a simple single construction of a `SecureRandom` object without any algorithm preferences. This will select `NativePRNG`, but is the preferred form of doing so. We explicitly don't want `SecureRandom.getInstanceStrong()` since it may block and may cause other apps to block for no cryptographic gain.
- Extract the codegolf-style preference keys into an enum.
- Renamed `AMBIGUOUS` to `NO_AMBIGUOUS` to at least get the triple negations consistent. Further renaming might be helpful.
- Get rid of the `NO_VOWELS` option for completely random passwords. It wasn't accessible from the UI and does not seem to be very useful compared to the `NO_AMBIGUOUS` option.
- Do not remove already chosen words from the XKCD word list. It needlessly complicates the code and (negligibly) reduces entropy for apparently no gain.

## :bulb: Motivation and Context

I found the string `SHA1PRNG` in our codebase and got (unnecessarily) scared. I then discovered that our password generation was pretty much incomprehensible. In the end, it didn't turn out to be broken though.

## :green_heart: How did you test it?
I generated passwords and compared with a current release. The types of generated passwords (characters that appear, phoneme structure) seem to match.

## :pencil: Checklist

- [x] I formatted the code with the IDE's reformat action (Ctrl + Shift + L/Cmd + Shift + L)
- [x] I reviewed submitted code
- [x] I added a [CHANGELOG](CHANGELOG.md) entry if applicable

### CODE REVIEW

**Comment by FabianHenneke** [comment]
I set the 1.10.0 milestone for this since I want it to get tested a bit more before releasing it.

File: app/src/main/java/com/zeapo/pwdstore/pwgen/PasswordGenerator.kt [hunk 1]
```diff
@@ -7,139 +7,131 @@ package com.zeapo.pwdstore.pwgen
...
+    }
+
     /**
-     * Generates passwords using the preferences set by
-     * [.setPrefs].
-     *
-     * @param ctx context from which to retrieve SharedPreferences from
-     * preferences file 'PasswordGenerator'
-     * @return list of generated passwords
+     * Generates a password using the preferences set by [.setPrefs].
```
Comment (msfjarvis) [inline app/src/main/java/com/zeapo/pwdstore/pwgen/PasswordGenerator.kt:65]:
```suggestion
     * Generates a password using the preferences set by [setPrefs].
```

I believe it works without the period

File: app/src/main/java/com/zeapo/pwdstore/pwgen/RandomPhonemesGenerator.kt [hunk 2]
```diff
@@ -0,0 +1,167 @@
...
+        Element("th", CONSONANT or DIPHTHONG),
+        Element("u", VOWEL),
+        Element("v", CONSONANT),
+        Element("w", CONSONANT),
+        Element("x", CONSONANT),
+        Element("y", CONSONANT),
+        Element("z", CONSONANT)
+    )
+
+    private class Element constructor(str: String, val flags: Int) {
```
Comment (msfjarvis) [inline app/src/main/java/com/zeapo/pwdstore/pwgen/RandomPhonemesGenerator.kt:59]: why is the constructor keyword required?

File: app/src/main/java/com/zeapo/pwdstore/pwgenxkpwd/PasswordBuilder.kt [hunk 3]
```diff
@@ -155,12 +136,10 @@ class PasswordBuilder(ctx: Context) {
         return password.toString()
     }
 
-    private fun capitalize(s: String): String {
-        var result = s
-        val lower = result.toLowerCase(Locale.getDefault())
-        result = lower.substring(0, 1).toUpperCase(Locale.getDefault()) + result.substring(1)
-        return result
-    }
+    private fun capitalize(s: String) =
```
Comment (msfjarvis) [inline app/src/main/java/com/zeapo/pwdstore/pwgenxkpwd/PasswordBuilder.kt:139]: can this be replaced with the stdlib capitalize method?

File: app/src/main/java/com/zeapo/pwdstore/ui/dialogs/PasswordGeneratorDialogFragment.kt [hunk 4]
```diff
@@ -4,101 +4,96 @@
...
-            setPrefs(requireActivity().applicationContext, preferences)
+            passwordField.text = generate(requireContext().applicationContext)
+        } catch (e: PasswordGeneratorException) {
+            Toast.makeText(requireActivity(), e.message, Toast.LENGTH_SHORT).show()
+            passwordField.text = ""
         }
     }
+
+    private fun isChecked(@IdRes id: Int): Boolean {
+        return dialog!!.findViewById<CheckBox>(id).isChecked
```
Comment (msfjarvis) [inline app/src/main/java/com/zeapo/pwdstore/ui/dialogs/PasswordGeneratorDialogFragment.kt:82]: there should be a `requireDialog()` method from AppCompat that is available here

File: app/src/main/java/com/zeapo/pwdstore/ui/dialogs/PasswordGeneratorDialogFragment.kt [hunk 5]
```diff
@@ -4,101 +4,96 @@
...
+    private fun setPreferences() {
+        val preferences = listOfNotNull(
+            PasswordOption.NoDigits.takeIf { !isChecked(R.id.numerals) },
+            PasswordOption.AtLeastOneSymbol.takeIf { isChecked(R.id.symbols) },
+            PasswordOption.NoUppercaseLetters.takeIf { !isChecked(R.id.uppercase) },
+            PasswordOption.NoAmbiguousCharacters.takeIf { !isChecked(R.id.ambiguous) },
+            PasswordOption.FullyRandom.takeIf { !isChecked(R.id.pronounceable) },
+            PasswordOption.NoLowercaseLetters.takeIf { !isChecked(R.id.lowercase) }
+        )
+        val lengthText = dialog!!.findViewById<EditText>(R.id.lengthNumber).text.toString()
```
Comment (msfjarvis) [inline app/src/main/java/com/zeapo/pwdstore/ui/dialogs/PasswordGeneratorDialogFragment.kt:94]: Same here

File: app/src/main/java/com/zeapo/pwdstore/pwgenxkpwd/PasswordBuilder.kt (hunk 3 above)
Comment (FabianHenneke) [reply app/src/main/java/com/zeapo/pwdstore/pwgenxkpwd/PasswordBuilder.kt:139]: That call cares about even more Unicode case mapping weirdness, so I believe the answer is yes. I do love the Kotlin stdlib.

**Comment by FabianHenneke** [comment]
@msfjarvis My `.idea/gradle.xml` is still changed on every build, with the `useQualifiedModuleNames` line being deleted. Do you know what you did to keep it in there? `Convert Module Groups to Qualified Names` in Android Studio does nothing.

**Comment by msfjarvis** [comment]
> @msfjarvis My `.idea/gradle.xml` is still changed on every build, with the `useQualifiedModuleNames` line being deleted. Do you know what you did to keep it in there? `Convert Module Groups to Qualified Names` in Android Studio does nothing.

That was precisely what I did as far as I remember, what version of Studio are you on?

**Comment by FabianHenneke** [comment]
> > @msfjarvis My `.idea/gradle.xml` is still changed on every build, with the `useQualifiedModuleNames` line being deleted. Do you know what you did to keep it in there? `Convert Module Groups to Qualified Names` in Android Studio does nothing.
> 
> That was precisely what I did as far as I remember, what version of Studio are you on?

4.0. I have unchecked the checkbox about remembering the mapping and invalidated my cached, but the line is still deleted on every IDE operation. I will try to reimport the project.

**Comment by FabianHenneke** [comment]
> > > @msfjarvis My `.idea/gradle.xml` is still changed on every build, with the `useQualifiedModuleNames` line being deleted. Do you know what you did to keep it in there? `Convert Module Groups to Qualified Names` in Android Studio does nothing.
> > 
> > 
> > That was precisely what I did as far as I remember, what version of Studio are you on?
> 
> 4.0. I have unchecked the checkbox about remembering the mapping and invalidated my cached, but the line is still deleted on every IDE operation. I will try to reimport the project.

Deep cleaning everything and reimporting also doesn't fix the problem. Which version are you using? Guess I'll just update to that.

**Comment by msfjarvis** [comment]
> > > > @msfjarvis My `.idea/gradle.xml` is still changed on every build, with the `useQualifiedModuleNames` line being deleted. Do you know what you did to keep it in there? `Convert Module Groups to Qualified Names` in Android Studio does nothing.
> > > 
> > > 
> > > That was precisely what I did as far as I remember, what version of Studio are you on?
> > 
> > 
> > 4.0. I have unchecked the checkbox about remembering the mapping and invalidated my cached, but the line is still deleted on every IDE operation. I will try to reimport the project.
> 
> Deep cleaning everything and reimporting also doesn't fix the problem. Which version are you using? Guess I'll just update to that.

I'm on 4.2.0 Canary 1 right now but I believe 4.1 should work for this too.

**Comment by FabianHenneke** [comment]
Updating to 4.1 fixed the issue.

I have also addressed all the comments, but the CI is flaky again.

**Comment by msfjarvis** [comment]
> Updating to 4.1 fixed the issue.
> 
> I have also addressed all the comments, but the CI is flaky again.

Yeah something is up with Actions today but I'm not seeing a reported outage on https://status.github.com/

**Comment by FabianHenneke** [comment]
The previous fail for this PR was because `curl` failed to reach `dl.google.com`. Maybe this is not a GitHub issue after all.

**Pull request merged**